Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/Gemfile.lock

```diff
@@ -0,0 +1,230 @@
+GEM
```

Comment:
what is this file for?

Reply:
`Gemfile.lock` file is created automatically when installing project dependencies (gems) either through `bundle install` or `gem install <gems>`. It's a snapshot of all the gems and its version used that worked when running the application.

Reply:
1. "Gemfile.lock file is created automatically when installing project dependencies (gems) either through bundle install or gem install <gems>" Wrong. read again.
2. So, whats the difference between this and `Gemfile`?

Reply:
1. `Gemfile.lock' is created automatically when installing project dependencies (gems) through `bundle install`. `gem install <gems>` installs the gems on the local machine that we are working on.
2. `Gemfile` is where we list out the gems we needed, then run `bundle install` to create a `Gemfile.lock` file so that when running the same project on another machine, it will look at the `Gemfile.lock` to know the versions for each gems the application works on rather than just getting the latest version of all the gems in the `Gemfile`.

Reply:
1. Good. Be more careful. Small details makes a complete difference in the meaning
2. Why do we need to know the version of each gem?

Reply:
2. Running `bundle install` when `Gemfile.lock` is present would ensures that it only installs using the dependencies listed in there. Having a "lock" to specific version of each gem ensures that during deploying code to production there would not be any changes to the version of each gem that would possibly causes breaks.